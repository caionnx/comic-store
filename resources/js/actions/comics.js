import marvelApi from '../api/marvelApi'

const parseResults = (data) => ({
  comics: data.results,
  offset: parseInt(data.offset, 10),
  count: parseInt(data.count, 10),
  total: parseInt(data.total, 10)
})

export const setComics = (comics) => ({
  type: 'SET_COMICS',
  comics
})

export const addComics = (comics) => ({
  type: 'ADD_COMICS',
  comics
})

export const startSetComics = (params) => (dispatch) => {
  return marvelApi.getComics(params).then((data) => {
    const { comics, offset, count, total } = parseResults(data)
    dispatch(setComics(comics))

    return { offset, count, total }
  })
}

export const startAddComics = (params) => (dispatch) => {
  return marvelApi.getComics(params).then((data) => {
    const { comics, offset, count, total } = parseResults(data)
    dispatch(addComics(comics))

    return { offset, count, total }
  })
}
