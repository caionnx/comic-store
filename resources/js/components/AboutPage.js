import React from 'react'
import RandExp from 'randexp'

const AboutPage = () => (
  <div className='l-content-container'>
    <h3>About</h3>

    <p>This project was inspired by a challenge of <a target='_blank' href='https://github.com/contabilizei/front-end-teste'>Contabilizei</a>.</p>

    <p>
      Use Marvel's Comic API: <a target='_blank' href='https://developer.marvel.com/'><strong>developer.marvel.com/</strong></a>.
      In addition, it generates 10% of the Comics as 'rare' item, defined by the top thick border.
    </p>

    <p>
      In the 'My Cart' page you can apply discount coupons. There are two types, for regular comic and for rare item.
      <br />
      Try this:
      <strong> {new RandExp(/[0-9]\w{3,5}([A-Z]{2})/).gen()}</strong> or <strong>{new RandExp(/([A-Z]{2})[0-9]{4}[a-z]/).gen()}</strong>
    </p>

    <p>~> <a target='_blank' href='https://github.com/caionnx/comic-store'>Github page</a>.</p>
  </div>
)

export default AboutPage
