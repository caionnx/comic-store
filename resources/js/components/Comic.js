import React from 'react'

class Comic extends React.Component {
  state = {
    imageFormat: 'portrait_medium'
  }
  render () {
    const image = this.props.comic.images[0]
    const imagePath = image ? image.path : false
    const imageExtension = image ? image.extension : false

    return (
      <div>
        <h3>{this.props.comic.title}</h3>
        { imagePath && imageExtension &&
          <img src={`${imagePath}/${this.state.imageFormat}.${imageExtension}`} />
        }
      </div>
    )
  }
}

export default Comic
