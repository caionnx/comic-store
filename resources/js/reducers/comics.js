export default (state = [], action) => {
  switch (action.type) {
    case 'SET_COMICS':
      return action.comics
    case 'ADD_COMICS':
      return [...state, ...action.comics]
    default:
      return state
  }
}
