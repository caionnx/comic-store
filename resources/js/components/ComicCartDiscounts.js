import React from 'react'
import { connect } from 'react-redux'
import { editComicOfCart } from '../actions/cart'

class ComicCartDiscounts extends React.Component {
  discounts = {
    applied: [],
    list: [
      {
        test: /[0-9]\w+([A-Z]{2})/,
        value: 10,
        type: 'normal',
        rule: (comic) => !comic.rareIssue
      },
      {
        test: /([A-Z]{2})[0-9]{4}[a-z]/,
        value: 5,
        type: 'rare',
        rule: (comic) => comic.rareIssue
      }
    ]
  }

  getTypeOfDiscount = (code) => {
    let toReturn = null
    for (const discount of this.discounts.list) {
      const { test, type, value, rule } = discount
      const match = code.match(test)
      if (match) {
        toReturn = { type, value, rule }
        break
      }
    }

    return toReturn
  }

  getPrice = (comic) => {
    const { price } = comic.prices.find(p => p.price !== 0) || { price: 0 }
    return price || 0
  }

  handleDiscount = (discount) => {
    const comicsToApply = this.props.cart.filter(discount.rule)

    comicsToApply.forEach(comic => {
      const price = this.getPrice(comic)
      const priceWithDiscount = price - ((price * discount.value) / 100)

      this.props.updateComic(comic.id, { priceWithDiscount })
    })
  }

  onApplyDiscount = (ev) => {
    ev.preventDefault()
    const { applied: appliedDiscounts } = this.discounts
    const code = ev.target.querySelector('#discount-input').value
    const typeOfDiscount = this.getTypeOfDiscount(code)
    if (!appliedDiscounts.includes(code) && typeOfDiscount) {
      this.handleDiscount(typeOfDiscount)
      appliedDiscounts.push(code)
    }
  }
  render () {
    return (
      <form onSubmit={this.onApplyDiscount} className='l-input-group'>
        <div className='l-input-group__item'>
          <input id='discount-input' className='c-input' type='text' placeholder='Insert code' />
        </div>
        <div className='l-input-group__item'>
          <button className='c-button c-button--full-width'>Apply discount</button>
        </div>
      </form>
    )
  }
}

const mapStateToProps = (state) => ({
  cart: state.cart
})

const mapDispatchToProps = (dispatch) => ({
  updateComic: (id, updates) => dispatch(editComicOfCart(id, updates))
})

export default connect(mapStateToProps, mapDispatchToProps)(ComicCartDiscounts)
