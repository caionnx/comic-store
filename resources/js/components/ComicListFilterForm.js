import React from 'react'
import { connect } from 'react-redux'
import { toggleFull } from '../actions/fetching'
import { setText, setSearchParams } from '../actions/filter'
import { startSetComics } from '../actions/comics'

class ComicListFilterForm extends React.Component {
  onFormSubmit = (e) => {
    e.preventDefault()
    const {
      filter,
      toggleFetchingFull,
      setFilterText,
      startSetComics,
      setSearchParams
    } = this.props
    const title = e.target.querySelector('#title-input').value
    const validParam = title.length > 0 ? { title } : filter.minimal

    toggleFetchingFull()
    setFilterText(title)

    startSetComics(validParam).then(data => {
      const { offset, count, total } = data
      this.props.setSearchParams({ offset, count, total })

      toggleFetchingFull()
    })
  }

  onClearForm = (e) => {
    e.preventDefault()
    const {
      setFilterText,
      startSetComics,
      setSearchParams,
      filter
    } = this.props
    const titleElem = e.target.closest('form').querySelector('#title-input')
    titleElem.value = ''

    setFilterText(null)
    startSetComics(filter.minimal).then(data => {
      const { offset, count, total } = data
      setSearchParams({ offset, count, total })
    })
  }

  componentWillUnmount () {
    const { setFilterText } = this.props
    setFilterText(null)
  }

  render () {
    return (
      <form onSubmit={this.onFormSubmit} className='l-input-group'>
        <div className='l-input-group__item'>
          <input className='c-input' type='text' id='title-input' placeholder='Insert comic title' />
        </div>
        <div className='l-input-group__item'>
          <button className='c-button c-button--full-width' disabled={this.props.fetching.full}>Search</button>
        </div>
        { this.props.filter.text &&
          <div className='l-input-group__item'>
            <button className='c-button c-button--full-width' onClick={this.onClearForm}>Clear</button>
          </div>
        }
      </form>
    )
  }
}

const mapStateToProps = (state) => ({
  filter: state.filter,
  fetching: state.fetching
})

const mapDispatchToProps = (dispatch) => ({
  toggleFetchingFull: () => dispatch(toggleFull()),
  setFilterText: (text) => dispatch(setText(text)),
  setSearchParams: (params) => dispatch(setSearchParams(params)),
  startSetComics: (params) => dispatch(startSetComics(params))
})

export default connect(mapStateToProps, mapDispatchToProps)(ComicListFilterForm)
