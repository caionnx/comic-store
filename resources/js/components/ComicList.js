import React from 'react'
import PropTypes from 'prop-types'
import Comic from './Comic'

export default class ComicList extends React.Component {
  render () {
    return (
      <div className={`c-comic-list ${this.props.toCartListView ? 'c-comic-list--rows' : ''}`}>
        { this.props.comics.map((comic, idx) =>
          <Comic comic={comic} toCartListView={this.props.toCartListView} key={idx} />)
        }
      </div>
    )
  }
}

ComicList.propTypes = {
  comics: PropTypes.array.isRequired,
  toCartListView: PropTypes.any
}
