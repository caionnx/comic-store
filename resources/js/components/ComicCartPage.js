import React from 'react'
import { connect } from 'react-redux'
import round from '../utils/round'
import PriceWithDiscountComponent from './PriceWithDiscount'
import ComicList from './ComicList'
import ComicCartDiscounts from './ComicCartDiscounts'

class ComicCartPage extends React.Component {
  getPrice = (comic) => {
    const { price } = comic.prices.find(p => p.price !== 0) || { price: 0 }
    return price
  }

  getTotalAmount = () =>
    this.props.cart.reduce((total, item) => {
      total += this.getPrice(item)
      return total
    }, 0)

  getTotalAmountIncludingDiscount = () =>
    this.props.cart.reduce((total, item) => {
      const withDiscount = item.priceWithDiscount
      total += withDiscount || this.getPrice(item)
      return total
    }, 0)

  someHasDiscount = () =>
    !!this.props.cart.find(c => !!c.priceWithDiscount)

  render () {
    const amount = this.getTotalAmount()
    const amountIncludingDiscount = this.getTotalAmountIncludingDiscount()

    return (
      <div className='l-content-container'>
        <ComicList comics={this.props.cart} toCartListView />
        <p>Total:
          { this.someHasDiscount()
            ? <PriceWithDiscountComponent oldValue={amount} newValue={amountIncludingDiscount} />
            : ` $ ${round(amountIncludingDiscount)}`
          }
        </p>
        <ComicCartDiscounts />
      </div>
    )
  }
}

const mapStateToProps = (state) => ({
  cart: state.cart
})

export default connect(mapStateToProps)(ComicCartPage)
