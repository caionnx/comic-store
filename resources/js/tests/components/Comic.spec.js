import React from 'react'
import { shallow } from 'enzyme'
import { Comic } from '../../components/Comic'
import { comics } from '../fixtures/comics'
jest.mock('react-toastify/dist/ReactToastify.css', () => '')

const fakeComic = comics[0]
const fakeEv = { stopPropagation: jest.fn() }
let componentProps
let wrapper

beforeEach(() => {
  componentProps = {
    cart: [],
    addComicToCart: jest.fn(),
    removeComicFromCart: jest.fn()
  }

  wrapper = shallow(<Comic {...componentProps} comic={fakeComic} />)
})

afterEach(() => {
  jest.clearAllMocks()
})

test('Should comic match with snapshot', () => {
  expect(wrapper).toMatchSnapshot()
})

test('Should have rare className', () => {
  wrapper.setProps({ comic: { ...fakeComic, rareIssue: true } })
  const element = wrapper.find('.c-comic-list__item')

  expect(element.hasClass('is-rare')).toBeTruthy()
})

test('Should add comic to cart', () => {
  const buttonSelector = `#container-${fakeComic.id} > button`
  wrapper.find(buttonSelector).simulate('click', fakeEv)
  wrapper.setProps({ cart: [fakeComic] })
  const button = wrapper.find(buttonSelector)

  expect(componentProps.addComicToCart).toHaveBeenCalled()
  expect(button.text()).toBe('Remove')
})

test('Should remove comic from cart', () => {
  const buttonSelector = `#container-${fakeComic.id} > button`
  wrapper.find(buttonSelector).simulate('click', fakeEv) // Add
  wrapper.setProps({ cart: [fakeComic] }) // Add
  wrapper.find(buttonSelector).simulate('click', { // Remove
    ...fakeEv, target: { getAttribute: () => fakeComic.id }
  })
  wrapper.setProps({ cart: [] }) // Remove
  const button = wrapper.find(buttonSelector)

  expect(componentProps.removeComicFromCart).toHaveBeenCalled()
  expect(button.text()).toBe('Add to cart')
})

test('Should open modal on click', () => {
  const spy = jest.spyOn(wrapper.instance(), 'handleOpenModal')
  const element = wrapper.find('.c-comic-list__item')
  element.simulate('click', fakeEv)

  expect(spy).toHaveBeenCalled()
  expect(wrapper.state('showModal')).toBeTruthy()
})

test('Should open modal on keypress', () => {
  const spy = jest.spyOn(wrapper.instance(), 'handleOpenModal')
  const element = wrapper.find('.c-comic-list__item')
  element.simulate('keypress', fakeEv)

  expect(spy).toHaveBeenCalled()
  expect(wrapper.state('showModal')).toBeTruthy()
})

test('Should close modal', () => {
  const spy = jest.spyOn(wrapper.instance(), 'handleCloseModal')
  const element = wrapper.find('.c-comic-list__item')
  element.simulate('click', fakeEv)
  const modalCloseBtn = wrapper.find('.c-modal #close-modal')
  modalCloseBtn.simulate('click', fakeEv)

  expect(spy).toHaveBeenCalled()
  expect(wrapper.state('showModal')).toBeFalsy()
})
