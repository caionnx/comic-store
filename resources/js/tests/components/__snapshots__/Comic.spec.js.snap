// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should comic match with snapshot 1`] = `
<div
  className="c-comic-list__item "
  onClick={[Function]}
  onKeyPress={[Function]}
  tabIndex="0"
>
  <LazyImage
    height={150}
    offset={50}
    once={true}
    src="http://i.annihil.us/u/prod/marvel/i/mg/9/40/5b0f16c49c83a/portrait_medium.jpg"
    title="Captain America (2017) #703"
  />
  <div
    className="c-comic-list__item-container"
    id="container-67933"
  >
    <span
      className="c-comic-list__item-price"
    >
      $ 
      3.99
    </span>
    <button
      className="c-button c-button--full-width is-yellow"
      id={67933}
      onClick={[Function]}
    >
      Add to cart
    </button>
  </div>
  <Modal
    appElement={null}
    ariaHideApp={true}
    bodyOpenClassName="ReactModal__Body--open"
    className="c-modal"
    closeTimeoutMS={200}
    contentLabel="Info about Captain America (2017) #703"
    isOpen={false}
    onRequestClose={[Function]}
    parentSelector={[Function]}
    portalClassName="ReactModalPortal"
    shouldCloseOnEsc={true}
    shouldCloseOnOverlayClick={true}
    shouldFocusAfterRender={true}
    shouldReturnFocusAfterClose={true}
  >
    <div
      className="c-modal__body"
    >
      <h2
        className="c-modal__title"
      >
        Captain America (2017) #703
      </h2>
      <p>
        Jack Rogers is a wanted man on a desperate search for a cure to his son’s fatal disease! And once it’s found, a legend will be reborn!
      </p>
      <div
        className="c-modal__options"
      >
        <a
          className="c-button"
          href="http://marvel.com/comics/issue/67933/captain_america_2017_703?utm_campaign=apiRef&utm_source=11eccfa59c07a22aff04caca647aa978"
          target="_blank"
        >
          More info
        </a>
        <button
          className="c-button  is-yellow"
          id={67933}
          onClick={[Function]}
        >
          Add to cart
        </button>
        <button
          className="c-button"
          id="close-modal"
          onClick={[Function]}
        >
          Close
        </button>
      </div>
    </div>
  </Modal>
</div>
`;
